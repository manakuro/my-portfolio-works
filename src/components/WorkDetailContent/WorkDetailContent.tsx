import * as React from 'react'
import WorkContentLoader from '@/workContentLoader'

import './WorkDetailContent.css'
import { Work } from '@/modules/home/reducer'
import { onlyUpdateForKeys, withHandlers } from 'recompose'
import compose from 'recompose/compose'

export interface WorkDetailContentProps {
  updateWorkContentImg: (workContentImg: string) => void
  targetWork: Work | null
}

export interface WorkDetailContentWithHandlers {
  handleIntersection: () => any
}

export function handleIntersection(
  entry: IntersectionObserverEntry,
  props: WorkDetailContentProps,
): void {
  if (entry.isIntersecting) {
    const src = entry.target.getAttribute('data-src')
    if (src) props.updateWorkContentImg(src)
  }
}

export const enhance = compose<
  WorkDetailContentProps & WorkDetailContentWithHandlers,
  WorkDetailContentProps
>(
  withHandlers({
    handleIntersection: (props: WorkDetailContentProps) => (
      entry: IntersectionObserverEntry,
    ) => handleIntersection(entry, props),
  }),
  onlyUpdateForKeys(['targetWork']),
)

export const WorkDetailContentComponent = (
  props: WorkDetailContentProps & WorkDetailContentWithHandlers,
): JSX.Element | null => {
  if (!props.targetWork) return null

  const Component = WorkContentLoader[props.targetWork.component]

  return (
    <div className="markdown-body">
      <Component handleIntersection={props.handleIntersection} />
    </div>
  )
}

export const WorkDetailContentComponentEnhanced = enhance(
  WorkDetailContentComponent,
)

export default class WorkDetailContent extends React.PureComponent<
  WorkDetailContentProps,
  {}
> {
  constructor(props: WorkDetailContentProps) {
    super(props)
  }

  public render(): JSX.Element {
    return <WorkDetailContentComponentEnhanced {...this.props} />
  }
}
