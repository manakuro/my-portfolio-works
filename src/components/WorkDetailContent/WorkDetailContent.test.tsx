import * as React from 'react'
import { shallow } from 'enzyme'
import * as WorkDetailContent from '@/components/WorkDetailContent/WorkDetailContent'

const propsData: WorkDetailContent.WorkDetailContentProps &
  WorkDetailContent.WorkDetailContentWithHandlers = {
  targetWork: {
    id: 1,
    title: 'EC Website',
    description: `
      Iâ€™ve been a CMT for ten years now, and I have, literally and
      figuratively, held the pulse of a steaming cross-section of
      San Franciscans
      `,
    img: 'http://sample.jpg',
    component: 'WorkOne',
    languages: [1, 2],
  },
  updateWorkContentImg: jest.fn(),
  handleIntersection: jest.fn(),
}

describe('WorkDetailContent', () => {
  let props: WorkDetailContent.WorkDetailContentProps &
    WorkDetailContent.WorkDetailContentWithHandlers
  beforeEach(() => {
    props = { ...propsData }
  })

  let wrapper

  it('should render', () => {
    wrapper = shallow(
      <WorkDetailContent.WorkDetailContentComponent {...props} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should return null when there is no target component', () => {
    props.targetWork = null

    wrapper = shallow(
      <WorkDetailContent.WorkDetailContentComponent {...props} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  describe('WorkDetailContent Class', () => {
    it('should render enhanced component', () => {
      wrapper = shallow(<WorkDetailContent.default {...props} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('handleIntersection', () => {
    it('should call updateWorkContentImg', () => {
      const entry: any = {
        isIntersecting: true,
        target: {
          getAttribute: jest.fn(() => ({ src: 'src!' })),
        },
      }
      props.updateWorkContentImg = jest.fn()

      WorkDetailContent.handleIntersection(entry, props)

      expect(props.updateWorkContentImg).toHaveBeenCalledWith({ src: 'src!' })
    })

    it('should NOT call updateWorkContentImg when there is no src', () => {
      const entry: any = {
        isIntersecting: true,
        target: {
          getAttribute: jest.fn(() => null),
        },
      }
      props.updateWorkContentImg = jest.fn()

      WorkDetailContent.handleIntersection(entry, props)

      expect(props.updateWorkContentImg).not.toHaveBeenCalled()
    })

    it('should NOT call updateWorkContentImg when isIntersecting is false', () => {
      const entry: any = {
        isIntersecting: false,
        target: {
          getAttribute: jest.fn(() => null),
        },
      }
      props.updateWorkContentImg = jest.fn()

      WorkDetailContent.handleIntersection(entry, props)

      expect(props.updateWorkContentImg).not.toHaveBeenCalled()
    })
  })
})
