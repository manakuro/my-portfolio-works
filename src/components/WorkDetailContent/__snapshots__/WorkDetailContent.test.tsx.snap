// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkDetailContent WorkDetailContent Class should render enhanced component 1`] = `
<withHandlers(onlyUpdateForKeys(Component))
  handleIntersection={[MockFunction]}
  targetWork={
    Object {
      "component": "WorkOne",
      "description": "
      Iâ€™ve been a CMT for ten years now, and I have, literally and
      figuratively, held the pulse of a steaming cross-section of
      San Franciscans
      ",
      "id": 1,
      "img": "http://sample.jpg",
      "languages": Array [
        1,
        2,
      ],
      "title": "EC Website",
    }
  }
  updateWorkContentImg={[MockFunction]}
/>
`;

exports[`WorkDetailContent should render 1`] = `
<div
  className="markdown-body"
>
  <LoadableComponent
    handleIntersection={[MockFunction]}
  />
</div>
`;

exports[`WorkDetailContent should return null when there is no target component 1`] = `""`;
